File: 2025/09 September/08 Merge Sort for Linked List/main.cpp
// Link: https://www.geeksforgeeks.org/problems/sort-a-linked-list/1

#include <iostream>
#include <vector>
using namespace std;

class Node
{
public:
    int data;
    Node *next;

    Node(int x)
    {
        data = x;
        next = nullptr;
    }
};

Node *createLinkedList(vector<int> arr)
{
    Node *head = new Node(arr[0]);
    Node *tail = head;

    for (int i = 1; i < arr.size(); i++)
    {
        tail->next = new Node(arr[i]);
        tail = tail->next;
    }

    return head;
}

void printLinkedList(Node *head)
{
    Node *temp = head;
    while (temp != NULL)
    {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

class Solution
{
public:
    Node *split(Node *head)
    {
        Node *fast = head;
        Node *slow = head;

        while (fast != nullptr && fast->next != nullptr)
        {
            fast = fast->next->next;
            if (fast != nullptr)
            {
                slow = slow->next;
            }
        }

        Node *temp = slow->next;
        slow->next = nullptr;
        return temp;
    }

    Node *merge(Node *first, Node *second)
    {
        if (first == nullptr)
            return second;
        if (second == nullptr)
            return first;

        if (first->data < second->data)
        {
            first->next = merge(first->next, second);
            return first;
        }
        else
        {
            second->next = merge(first, second->next);
            return second;
        }
    }

    Node *mergeSort(Node *head)
    {
        if (head == nullptr || head->next == nullptr)
            return head;

        Node *second = split(head);

        head = mergeSort(head);
        second = mergeSort(second);

        return merge(head, second);
    }
};

int main()
{
    Solution sol;

    vector<int> arr = {40, 20, 60, 10, 50, 30};
    Node *head = createLinkedList(arr);
    cout << "Original Linked List: " << endl;
    printLinkedList(head);
    head = sol.mergeSort(head);
    cout << "Sorted Linked List: " << endl;
    printLinkedList(head);
    cout << endl;

    arr = {9, 5, 2, 8};
    head = createLinkedList(arr);
    cout << "Original Linked List: " << endl;
    printLinkedList(head);
    head = sol.mergeSort(head);
    cout << "Sorted Linked List: " << endl;
    printLinkedList(head);
    cout << endl;

    return 0;
}

File: 2025/09 September/08 Merge Sort for Linked List/main.py
# Link: https://www.geeksforgeeks.org/problems/sort-a-linked-list/1


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def createLinkedList(arr):
    head = Node(arr[0])
    temp = head
    for i in range(1, len(arr)):
        temp.next = Node(arr[i])
        temp = temp.next
    return head


def printLinkedList(head):
    temp = head
    while temp is not None:
        print(temp.data, end=" ")
        temp = temp.next
    print()


class Solution:
    def split(self, head):
        fast = head
        slow = head

        while fast and fast.next:
            fast = fast.next.next
            if fast:
                slow = slow.next

        second = slow.next
        slow.next = None
        return second

    def merge(self, first, second):
        if not first:
            return second
        if not second:
            return first

        if first.data < second.data:
            first.next = self.merge(first.next, second)
            return first
        else:
            second.next = self.merge(first, second.next)
            return second

    def mergeSort(self, head):
        if not head or not head.next:
            return head

        second = self.split(head)

        head = self.mergeSort(head)
        second = self.mergeSort(second)

        return self.merge(head, second)


if __name__ == "__main__":
    solution = Solution()

    arr = [40, 20, 60, 10, 50, 30]
    head = createLinkedList(arr)
    print("Original Linked List:")
    printLinkedList(head)
    sorted_head = solution.mergeSort(head)
    print("Sorted Linked List:")
    printLinkedList(sorted_head)
    print()

    arr = [9, 5, 2, 8]
    head = createLinkedList(arr)
    print("Original Linked List:")
    printLinkedList(head)
    sorted_head = solution.mergeSort(head)
    print("Sorted Linked List:")
    printLinkedList(sorted_head)
    print()
